File: test/render/test-cases/data-filter-extension.js
import {ScatterplotLayer} from '@deck.gl/layers';
import {DataFilterExtension} from '@deck.gl/extensions';

const VIEWSTATE = {latitude: 37, longitude: -122, zoom: 10.2, pitch: 0, bearing: 0};
const DATA = [];
let i = 0;
for (let y = 0; y < 20; y++) {
  for (let x = 0; x < 32; x++) {
    DATA.push({
      position: [VIEWSTATE.longitude + 0.01 * (x - 16), VIEWSTATE.latitude - 0.01 * (y - 10)],
      index: i++,
      column: x,
      row: y
    });
  }
}

export default [
  {
    name: 'filter-1d',
    props: {
      extensions: [new DataFilterExtension({filterSize: 1})],
      getFilterValue: d => d.column,
      filterRange: [3, 8]
    }
  },
  {
    name: 'filter-2d',
    props: {
      extensions: [new DataFilterExtension({filterSize: 2})],
      getFilterValue: d => [d.column, d.row],
      filterRange: [
        [3, 8],
        [2, 5]
      ]
    }
  },
  {
    name: 'single-category',
    props: {
      extensions: [new DataFilterExtension({categorySize: 1})],
      getFilterCategory: d => d.index % 128,
      filterCategories: [3, 4, 5, 127, 0]
    }
  },
  {
    name: '2-categories',
    props: {
      extensions: [new DataFilterExtension({categorySize: 2, filterSize: 0})],
      getFilterCategory: d => [d.index % 128, d.row],
      filterCategories: [
        [3, 4, 5, 15, 50],
        [4, 5, 8, 16, 'missing']
      ]
    }
  },
  {
    name: '3-categories',
    props: {
      extensions: [new DataFilterExtension({categorySize: 3})],
      getFilterCategory: d => [d.index % 128, d.column, d.row],
      filterCategories: [[3], [1, 2, 3, 4, 5], [0, 1, 2, 3, 4, 5, 6]]
    }
  },
  {
    name: '4-categories',
    props: {
      extensions: [new DataFilterExtension({categorySize: 4})],
      getFilterCategory: ({index}) => [index % 2, index % 3, index % 5, index % 7],
      filterCategories: [[0], [0], [0], [0]]
    }
  },
  {
    name: '4-categories-filter-2d',
    props: {
      extensions: [new DataFilterExtension({categorySize: 4, filterSize: 2})],
      getFilterCategory: ({index}) => [index % 2, index % 3, index % 5, index % 7],
      filterCategories: [[0], [0], [0], [0]],
      getFilterValue: d => [d.column, d.row],
      filterRange: [
        [2, 1000],
        [0, 15]
      ]
    }
  }
].map(({name, props}) => ({
  name: `data-filter-effect-${name}`,
  viewState: VIEWSTATE,
  layers: [
    new ScatterplotLayer({
      id: name,
      data: DATA,
      getRadius: 8,
      radiusUnits: 'pixels',
      ...props
    }),
    new ScatterplotLayer({
      id: `${name}-background`,
      data: DATA,
      getRadius: 2,
      radiusUnits: 'pixels'
    })
  ],
  goldenImage: `./test/render/golden-images/data-filter-effect-${name}.png`
}));

File: bindings/pydeck/examples/data_filter_extension.py
"""
DataFilterExtension
===================

Plot of BART (Bay Area Rapid Transit) stations in San Francisco, filtered on the GPU by
number of exits using the deck.gl ``DataFilterExtension``.

Only stations whose ``exits`` value falls within ``filter_range`` are drawn. Extension
options (here ``filter_size``) are passed to ``pdk.Extension``, while the props the
extension adds to the layer (``get_filter_value`` and ``filter_range``) are passed to the
``pdk.Layer``.

Adapted from the ScatterplotLayer example.
"""

import pydeck as pdk
import pandas as pd
import math

DATA_URL = "https://raw.githubusercontent.com/visgl/deck.gl-data/master/website/bart-stations.json"
df = pd.read_json(DATA_URL)

# Use pandas to calculate additional data
df["exits_radius"] = df["exits"].apply(lambda exits_count: math.sqrt(exits_count))

# Only show stations with at least this many exits
MIN_EXITS = 5000

layer = pdk.Layer(
    "ScatterplotLayer",
    df,
    pickable=True,
    opacity=0.8,
    stroked=True,
    filled=True,
    radius_scale=6,
    radius_min_pixels=1,
    radius_max_pixels=100,
    line_width_min_pixels=1,
    get_position="coordinates",
    get_radius="exits_radius",
    get_fill_color=[255, 140, 0],
    get_line_color=[0, 0, 0],
    # Props added to the layer by the DataFilterExtension:
    get_filter_value="exits",
    filter_range=[MIN_EXITS, 100000],
    # The extension itself, with its own options:
    extensions=[pdk.Extension("DataFilterExtension", filter_size=1)],
)

# Set the viewport location
view_state = pdk.ViewState(latitude=37.7749295, longitude=-122.4194155, zoom=10, bearing=0, pitch=0)

# Render
r = pdk.Deck(layers=[layer], initial_view_state=view_state, tooltip={"text": "{name}\n{address}"})
r.to_html("data_filter_extension.html")
